### ISSUE TITLE
Building on legacy chipsets

### ISSUE THREAD

**Issue by andymarden** [tigerbeetle/tigerbeetle#2406]
Raising as discussed on slack channel - few issues together and some are more general.
Firstly, I could not build the tigerbeetle server but got it working in the end by:

1. Changing the build.zig script to replace the cpus const with:

```
    const cpus = .{
        "baseline+aes+neon",
        "baseline+aes+neon",
        "westmere+aes"
        "x86_64_v3+aes",
        "x86_64_v3+aes",
    };
```
(the westmere line is the important one)

I also found the part in src/vsr/checksum.zig that says:

```
comptime {
    // As described above, TigerBeetle uses Aegis (and thus AES Blocks), for its checksumming.
    // While there is a software implementation, it's much slower and we don't expect to ever be
    // using it considering we target platforms with AES hardware acceleration.
    //
    // If you're trying to compile TigerBeetle for an older CPU without AES hardware acceleration,
    // you'll need to disable the following assert.
    assert(std.crypto.core.aes.has_hardware_support);
}
```
and commented out that line

Then I had a working server.

2. The c client would build but crashed at runtime. My CPUs are Xeon X5650's. On the slack channel, I was pointed at this issue as possible an indication of the problem with the client. It turns out that in build.zig, there is another set of cpu architectures which I changed to:

```
const platforms = .{
.{ "x86_64-linux-gnu.2.27", "linux-x64", "baseline" },
.{ "x86_64-linux-musl", "linux-musl-x64", "x86_64_v3+aes" },
.{ "x86_64-macos", "osx-x64", "x86_64_v3+aes" },
.{ "aarch64-linux-gnu.2.27", "linux-arm64", "baseline+aes+neon" },
.{ "aarch64-linux-musl", "linux-musl-arm64", "baseline+aes+neon" },
.{ "aarch64-macos", "osx-arm64", "baseline+aes+neon" },
.{ "x86_64-windows", "win-x64", "x86_64_v3+aes" },
};
```
(first line)

In that issue, it also notes that you can run build.zig with a -mcpus= flag but that does not work in the latest version.

So what do I think could be done to improve this experience:

1. Put the -mcpus flag back on and document its usage (both server and client builds)
2. Set the correct arch and cpu features based on what is passed in but default to what is in the script if not - you could even have a warning if it is overriden saying "we haven't tested on this so on your own head be it"
3. make sure there is no editing of scripts necessary to do a build - it should all be driven off the env and what is passed in

I don't actually know if what I did is correct but it works.

**Comment by batiati** [comment]
Hey @andymarden!
Thanks for documenting the changes required to build TigerBeetle on legacy/unsupported targets.

Currently, we don't plan to support software-based hash, so the targets (both replicas and clients) are required to support hardware-accelerated AES encryption.  As it heavily penalizes performance and increases the surface area for tests, we also don't provide official "patches" or build flags to relax the minimal CPU requirements.

It's possible that we will revisit this decision in the future (especially for clients) by implementing some kind of fallback algorithm for low-end devices.

**Comment by batiati** [comment]
Closed by #2407

**Comment by andymarden** [comment]
I believe, but that the first has settled, Westmere chips do support aes - I have it working anyhow. Point is that having docs out of line with what the code can do and having comments about commenting things out in the code isn't ideal. Perhaps I didn't follow the right path, but the journey was very painful. Up to you if you want to address that.
